File: erlang/etudes/ch6/stats.erl
-module(stats).
-export([minimum/1,maximum/1,range/1]).

minimum([Head|[]]) -> Head;
minimum([Head|Tail]) ->
    Min = minimum(Tail),
    if
        Head < Min -> Head;
        true -> Min
    end.

maximum([Head|[]]) -> Head;
maximum([Head|Tail]) ->
    Max = maximum(Tail),
    if
        Head > Max -> Head;
        true -> Max
    end.

range([Head|Tail]) -> range(Tail, Head, Head).
range([Head|Tail],Lower,Upper) when Head < Lower -> range(Tail, Head, Upper);
range([Head|Tail],Lower,Upper) when Head > Upper -> range(Tail, Lower, Head);
range([_|Tail],Lower,Upper) -> range(Tail, Lower, Upper);
range([],Lower,Upper) -> [Lower, Upper].

File: elixir/etudes/ch06/stats.ex
defmodule Stats do
  def minimum([x|[]]), do: x
  def minimum([x|tail]) do
    min = minimum(tail)
    if x > min, do: min, else: x
  end

  def maximum([x|[]]), do: x
  def maximum([x|tail]) do
    max = maximum(tail)
    if x > max, do: x, else: max
  end

  def range(a, b) when a == b, do: [a]
  def range(a, b), do: [a | range(a + 1, b)]
end
